src/pointarray.cpp

```diff
@@ -613,17 +685,58 @@ DrawCount PointArray::drawPoints(QGLShaderProgram& prog, const TransformState& t
         if (!incrementalDraw)
             node->nextBeginIndex = node->beginIndex;
 
+        // tfm::printfln("OCTREE NODE (STACK) begin idx: %i", idx);
+
         DrawCount nodeDrawCount = node->drawCount(relCamera, quality, incrementalDraw);
         drawCount += nodeDrawCount;
 
         idx = node->nextBeginIndex;
         if (nodeDrawCount.numVertices == 0)
             continue;
+
+        if (m_fields.size() < 1)
+            continue;
+
+        unsigned int bufferSize = 0;
+        for (size_t i = 0; i < m_fields.size(); ++i)
+        {
+            const GeomField &field = m_fields[i];
+            int arraySize = field.spec.arraySize();
+            int vecSize = field.spec.vectorSize();
+
+            // tfm::printfln("FIELD-NAME: %s", field.name);
+            // tfm::printfln("AS: %i, VS: %i, FSS: %i, FSES: %i, GLBTFSS: %i", arraySize, vecSize, field.spec.size(), field.spec.elsize, sizeof(glBaseType(field.spec)));
+
+            bufferSize += arraySize * vecSize * field.spec.elsize; //sizeof(glBaseType(field.spec));
+        }
+
+        bufferSize = bufferSize * (GLsizei)nodeDrawCount.numVertices;
+
+        // TODO: might be able to do something more efficient here, for example use glBufferSubData to avoid re-allocation of memory by glBufferData
+        // INITIALIZE THE BUFFER TO FULL SIZE
+        // tfm::printfln("INIT BUFFER: %i, BS: %i", vbo, bufferSize);
+
+        glBufferData(GL_ARRAY_BUFFER, bufferSize, NULL, GL_STREAM_DRAW);
+        /// ========================================================================
+        /// ========================================================================
+        int bufferOffset = 0;
+
         for (size_t i = 0, k = 0; i < m_fields.size(); k+=m_fields[i].spec.arraySize(), ++i)
         {
             const GeomField& field = m_fields[i];
             int arraySize = field.spec.arraySize();
             int vecSize = field.spec.vectorSize();
+
+            // TODO: should use a single data-array that isn't split into vertex / normal / color / etc. sections, but has interleaved data
+            // OpenGL has a stride value in glVertexAttribPointer for exactly this purpose, which should be used for better efficiency
+            // here we write only the current attribute data into this the buffer (e.g. all positions, then all colors)
```

Comment:
It was more of a suggestion, but I can see how this could make things more problematic on the CPU side. I think the only expected performance benefit would come from reducing the number of calls to `glDrawArrays`, which would also mean a reduced number of calls to `glBufferData`, but with larger buffer sizes, this might not be possible depending on how big that buffer is really going to get?